### PR TITLE
storagenode/pieces: remove buffering from reading/writing and fix io.EOF bug

### PR DESCRIPTION
**Pull request by thepaul**
What: 

* Remove `bufio.Reader` and `bufio.Writer` doesn't seem to have significant impact on reading/writing for large chunk sizes.
* Special case `io.EOF` since most Go code assumes them not to be wrapped.
* Fix bug in endpoint, which assumed `Read` always returns sufficient amount of data.

Please describe the tests:
    (No new functionality; existing tests should cover the changed behavior sufficiently.)
 
Please describe the performance impact:
    Most likely a performance improvement, due to using a read buffer (assuming we do some small or unaligned reads sometimes). Could potentially cause a slight slowdown for read operations in pathological cases, though.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

**Comment by thepaul** [comment]
wow ok I guess it does cause problems. should we take out the bufio.Reader entirely? or is there something dumb i'm missing?

**Comment by egonelbre** [comment]
Two things:
1. we probably should test that the bufio is useful
2. we should figure out what's causing the failure, since adding a buffer somewhere shouldn't break things.

**Comment by egonelbre** [comment]
PS: One common thing that breaks with adding buffering is wrong code that looks like:
```
n, err := buf.Read(data)
if err != nil {
	return err
}

process(data[:n])
```

Whereas, it should be:

```
n, err := buf.Read(data)
if n > 0 {
	process(data[:n])
}
if err != nil {
	return err
}
```

**Comment by thepaul** [comment]
Made a ticket to investigate further at V3-2158.

**Comment by egonelbre** [comment]
So I found a few suspicious read handling:

* mobile/bucket.go:333: resets n to -1 when reader returns io.EOF
* pkg/ranger/file.go:70: resetting of the error code looks suspicious
* storagenode/piecestore/endpoint.go:451: piecereader.Read isn't handling `n`

**Comment by egonelbre** [comment]
It seems the buffering doesn't matter when writing/reading 512KB at a time:

```
BEFORE:
$ go test -bench . -run XXXX ./storagenode/pieces                       
pkg: storj.io/storj/storagenode/pieces                                  
BenchmarkReadWrite/Write-32                    9         121431378 ns/op
BenchmarkReadWrite/Read-32                   184           6444910 ns/op
PASS                                                                    
ok      storj.io/storj/storagenode/pieces       3.423s                  

AFTER:
$ go test -bench . -run XXXX ./storagenode/pieces                       
goos: windows                                                           
goarch: amd64                                                           
pkg: storj.io/storj/storagenode/pieces                                  
BenchmarkReadWrite/Write-32                    9         126349222 ns/op
BenchmarkReadWrite/Read-32                   193           6145245 ns/op
```

File: storagenode/pieces/readwrite_test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	defer ctx.Check(blobs.Close)
+
+	store := pieces.NewStore(zap.NewNop(), blobs)
+
+	// setup test parameters
+	const blockSize = int(256 * memory.KiB)
+	satelliteID := testrand.NodeID()
+	source := testrand.Bytes(30 * memory.MB)
+
+	testPieceID := storj.PieceID{1}
```
Comment (mniewrzal) [inline storagenode/pieces/readwrite_test.go]: can you move writing a test piece after `Write` benchmark to be clear that its used only to test `Read`?

Comment (egonelbre) [reply storagenode/pieces/readwrite_test.go]: ah, just moved it before... but sure.

**Comment by egonelbre** [comment]
The grpc buffer sizes seem appropriate.

**Comment by mniewrzal** [comment]
hmm its 32KB according to godoc and our buffer was 256KiB before removing.

**Comment by egonelbre** [comment]
The grpc buffer size is different from the chunk size we transfer.

**Pull request merged**